cmd/algokey/keyreg.go

```diff
@@ -0,0 +1,254 @@
+	data, err := base64.StdEncoding.DecodeString(b64)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to decode digest '%s': %s\n\n", b64, err)
+		os.Exit(1)
+	}
+	if len(data) != len(digest[:]) {
+		fmt.Fprintf(os.Stderr, "Unexpected decoded digest length decoding '%s'.\n\n", b64)
+		os.Exit(1)
+	}
+	copy(digest[:], data)
+	return
+}
+
+func getGenesisInformation(network string) (crypto.Digest, error) {
+	// For testing purposes, there is a secret option to override the genesis information.
+	hashOverride := os.Getenv("ALGOKEY_GENESIS_HASH")
+	if hashOverride != "" {
+		return mustConvertB64ToDigest(hashOverride), nil
+	}
+
+	// Otherwise check that network matches one of the known networks.
+	gen, ok := validNetworks[strings.ToLower(network)]
+	if !ok {
+		return crypto.Digest{}, fmt.Errorf("unknown network '%s' provided. Supported networks: %s",
+			network,
+			strings.Join(validNetworkList, ", "))
+	}
+
+	return gen, nil
+}
+
+func run(params keyregCmdParams) error {
+	// Implicit last valid
+	if params.lastValid == 0 {
+		params.lastValid = params.firstValid + txnLife
+	}
+
+	if !params.offline {
+		if params.partkeyFile == "" {
+			return errors.New("must provide --keyfile when registering participation keys")
+		}
+		if params.addr != "" {
+			return errors.New("do not provide --address when registering participation keys")
+		}
+	} else {
+		if params.addr == "" {
+			return errors.New("must provide --address when bringing an account offline")
+		}
+		if params.partkeyFile != "" {
+			return errors.New("do not provide --keyfile when bringing an account offline")
+		}
+	}
+
+	var accountAddress basics.Address
+	if params.addr != "" {
+		var err error
+		accountAddress, err = basics.UnmarshalChecksumAddress(params.addr)
+		if err != nil {
+			return fmt.Errorf("unable to parse --address: %w", err)
+		}
+	}
+
+	if params.partkeyFile != "" && !util.FileExists(params.partkeyFile) {
+		return fmt.Errorf("cannot access keyfile '%s'", params.partkeyFile)
+	}
+
+	if params.txFile == "" {
+		params.txFile = fmt.Sprintf("%s.tx", params.partkeyFile)
+	}
+
+	if util.FileExists(params.txFile) || params.txFile == stdoutFilenameValue {
+		return fmt.Errorf("outputFile '%s' already exists", params.partkeyFile)
+	}
+
+	// Lookup information from partkey file
+	var part *account.Participation
+	if params.partkeyFile != "" {
+		partDB, err := db.MakeErasableAccessor(params.partkeyFile)
+		if err != nil {
+			return fmt.Errorf("cannot open keyfile %s: %v", params.partkeyFile, err)
+		}
+
+		partkey, err := account.RestoreParticipation(partDB)
+		if err != nil {
+			return fmt.Errorf("cannot load keyfile %s: %v", params.partkeyFile, err)
+		}
+		defer partkey.Close()
+
+		part = &partkey.Participation
+
+		if params.firstValid < uint64(part.FirstValid) {
+			return fmt.Errorf("firstvalid (%d) should be greater than or equal to the first valid participation key (%d)", params.firstValid, part.FirstValid)
```

Comment:
The poor non technical end user would not be able to understand from this what the issue is. This would not be an issue if this was an internal error, but given that it's meant for UI application, you might want to consider the following :
"generating a transactions requires that the transaction's firstValid field would be greater than the participation key's first valid round. The participation key you have provided has a first valid round of %d while the requested transaction firstvalid round is %d".

Reply:
We're not talking about an internal parameter that the user was unaware of. They were required to provide a few parameters and this message explains why it was rejected.

Reply:
Suggested wording:
"The transaction's firstvalid round (%d) field should be set greater than or equal to the participation key's first valid round (%d). The network will reject keyreg transactions that are set to take effect before the participation key's 
 first valid round."